docs/internals/layout_in_memory.rst

```diff
@@ -1,40 +1,123 @@
 
 .. index: memory layout
 
-****************
-Layout in Memory
-****************
+***********************
+Reserved Memory Areas
+***********************
 
 Solidity reserves four 32-byte slots, with specific byte ranges (inclusive of endpoints) being used as follows:
 
 - ``0x00`` - ``0x3f`` (64 bytes): scratch space for hashing methods
-- ``0x40`` - ``0x5f`` (32 bytes): currently allocated memory size (aka. free memory pointer)
+- ``0x40`` - ``0x5f`` (32 bytes): allocated memory size (aka. free memory pointer)
 - ``0x60`` - ``0x7f`` (32 bytes): zero slot
 
-Scratch space can be used between statements (i.e. within inline assembly). The zero slot
-is used as initial value for dynamic memory arrays and should never be written to
-(the free memory pointer points to ``0x80`` initially).
+Every time you access memory, the EVM knows the location that was accessed. Roughly speaking, the
+maximum location you access determines the size of active memory. Increasing the size of
+active memory increases your gas costs non-linearly. The specifics are beyond the scope
+of this documentation. More information about gas costs can be found in the Etherum Yellow Paper
+and in the EVM (go-ethereum) project and specifications.
+
+Solidity is a high level language that is translated (compiled) into EVM instructions. Its features
+work because of the conventions Solidity defines and uses, in combination with the implementations
+of complex data structures that it provides (such as multi-dimensional arrays). These data
+structures require that memory be allocated (reserved) for their use, so that instructions that
+execute after they're allocated don't corrupt them.
+
+Memory allocation is a high-level concept, not an EVM concept. Its implementation in Solidity relies
+on the value stored at memory location ``0x40``. That value tracks the next unallocated memory
```

Comment:
I'd probably use free memory pointer rather than `0x40`. And potentially document that it's currently at `0x40`.